core/src/test/java/io/grpc/internal/PickFirstLeafLoadBalancerTest.java

```diff
@@ -0,0 +1,2577 @@
+    assertThat(argsList.get(3).getAddresses().get(0)).isEqualTo(servers.get(3));
+    assertThat(argsList.get(0).getAddresses().size()).isEqualTo(1);
+    assertThat(argsList.get(1).getAddresses().size()).isEqualTo(1);
+    assertThat(argsList.get(2).getAddresses().size()).isEqualTo(1);
+    assertThat(argsList.get(3).getAddresses().size()).isEqualTo(1);
+    assertThat(argsList.get(0).getAttributes()).isEqualTo(Attributes.EMPTY);
+    assertThat(argsList.get(1).getAttributes()).isEqualTo(Attributes.EMPTY);
+    assertThat(argsList.get(2).getAttributes()).isEqualTo(Attributes.EMPTY);
+    assertThat(argsList.get(3).getAttributes()).isEqualTo(Attributes.EMPTY);
+
+    inOrder.verify(mockHelper).updateBalancingState(eq(CONNECTING), pickerCaptor.capture());
+    verify(mockSubchannel1).requestConnection();
+
+    assertNull(pickerCaptor.getValue().pickSubchannel(mockArgs)
+        .getSubchannel());
+
+    assertEquals(pickerCaptor.getValue().pickSubchannel(mockArgs),
+        pickerCaptor.getValue().pickSubchannel(mockArgs));
+
+    verifyNoMoreInteractions(mockHelper);
+  }
+
+  @Test
+  public void nameResolutionErrorWithStateChanges() throws Exception {
+    List<EquivalentAddressGroup> newServers = Lists.newArrayList(servers.get(0));
+    InOrder inOrder = inOrder(mockHelper);
+    loadBalancer.acceptResolvedAddresses(
+        ResolvedAddresses.newBuilder().setAddresses(newServers).setAttributes(affinity).build());
+    verify(mockHelper).createSubchannel(createArgsCaptor.capture());
+    verify(mockSubchannel1).start(stateListenerCaptor.capture());
+    inOrder.verify(mockHelper).updateBalancingState(eq(CONNECTING), any(SubchannelPicker.class));
+    SubchannelStateListener stateListener = stateListenerCaptor.getValue();
+
+    stateListener.onSubchannelState(ConnectivityStateInfo.forTransientFailure(Status.UNAVAILABLE));
+    inOrder.verify(mockHelper).refreshNameResolution();
+    inOrder.verify(mockHelper).updateBalancingState(
+        eq(TRANSIENT_FAILURE), any(SubchannelPicker.class));
+    Status error = Status.NOT_FOUND.withDescription("nameResolutionError");
+    loadBalancer.handleNameResolutionError(error);
+    inOrder.verify(mockHelper).updateBalancingState(eq(TRANSIENT_FAILURE), pickerCaptor.capture());
+    PickResult pickResult = pickerCaptor.getValue().pickSubchannel(mockArgs);
+    assertNull(pickResult.getSubchannel());
+    assertEquals(error, pickResult.getStatus());
+
+    Status error2 = Status.NOT_FOUND.withDescription("nameResolutionError2");
+    loadBalancer.handleNameResolutionError(error2);
+    inOrder.verify(mockHelper).updateBalancingState(eq(TRANSIENT_FAILURE), pickerCaptor.capture());
+
+    pickResult = pickerCaptor.getValue().pickSubchannel(mockArgs);
+    assertNull(pickResult.getSubchannel());
+    assertEquals(error2, pickResult.getStatus());
+
+    verifyNoMoreInteractions(mockHelper);
+  }
+
+  @Test
+  public void requestConnection() {
```

Comment:
Maybe more scenarios of subchannel states when doing `requestConnection` covered in other tests. But I was expecting we can unit test this function thoroughly here.

Reply:
What do you mean by unit test for this? Wouldn't verifying that the right subchannel is started when requestConnection is called enough?

Reply:
`requestconnection()` method is recursive, and depending on previous subchannels' states in the LB, behaviour is different. But this test case only covers one basic happy case. In fact, this setup is probably mostly covered by other test cases. 
I think it is faire to cover more scenarios for the `requestConnection` method's behaviour specifically.

Reply:
Ah, I see what you mean. I thought we test this but we do not test it by calling `requestConnection()` directly. I added some unit tests for this; let me know what you think.